api/src/main/java/org/openmrs/module/patientqueueing/api/impl/PatientQueueingServiceImpl.java

```diff
@@ -12,70 +12,130 @@
+     */
+    public List<PatientQueue> getPatientQueueList(Provider provider, Date fromDate, Date toDate, Location locationTo, Location locationFrom, Patient patient, PatientQueue.Status status) {
+        return dao.getPatientQueueList(provider, fromDate, toDate, locationTo, locationFrom, patient, status);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#completePatientQueue(org.openmrs.module.patientqueueing.model.PatientQueue)
+     */
+    @Override
+    public PatientQueue completePatientQueue(PatientQueue patientQueue) {
+        patientQueue.setStatus(PatientQueue.Status.COMPLETED);
+        return dao.savePatientQueue(patientQueue);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getIncompletePatientQueue(org.openmrs.Patient,
+     * org.openmrs.Location)
+     */
+    public PatientQueue getIncompletePatientQueue(Patient patient, Location locationTo) {
+
+        return dao.getIncompletePatientQueue(patient, locationTo);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#getMostRecentQueue(org.openmrs.Patient)
+     */
+    @Override
+    public PatientQueue getMostRecentQueue(Patient patient) {
+        return dao.getMostRecentQueue(patient);
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#assignVisitNumber(org.openmrs.module.patientqueueing.model.PatientQueue)
+     */
+    @Override
+    public PatientQueue assignVisitNumber(PatientQueue patientQueue) {
+        Date today = new Date();
+        List<PatientQueue> patientQueueList = getPatientQueueList(null, OpenmrsUtil.firstSecondOfDay(today), OpenmrsUtil.getLastMomentOfDay(today), null, null, patientQueue.getPatient(), null);
+
+        if (!patientQueueList.isEmpty() && patientQueueList.get(0).getVisitNumber() != null) {
+            patientQueue.setVisitNumber(patientQueueList.get(0).getVisitNumber());
+        } else {
+            patientQueue.setVisitNumber(generateVisitNumber(patientQueue.getLocationFrom(), patientQueue.getPatient()));
+        }
+        return patientQueue;
+    }
+
+    /**
+     * @see org.openmrs.module.patientqueueing.api.PatientQueueingService#generateVisitNumber(org.openmrs.Location,
+     * org.openmrs.Patient)
+     */
+    public String generateVisitNumber(Location location, Patient patient) {
+
+        Date today = new Date();
+
+        SimpleDateFormat formatterExt = new SimpleDateFormat(Context.getAdministrationService().getGlobalProperty("patientqueueing.defaultDateFormat"));
+
+        List<PatientQueue> patientQueues = getPatientQueueList(null, OpenmrsUtil.firstSecondOfDay(today), OpenmrsUtil.getLastMomentOfDay(today), null, location, null, null);
+
+        int nextNumberInQueue = 1;
+        if (!patientQueues.isEmpty()) {
+            String currentVisitNumber = patientQueues.get(0).getVisitNumber();
```

Comment:
Because am interest in any that meet the condition that are used to return the list as long as it it not empty. The first in list will give me what I need